src/zone/napa-zone.cpp

```diff
@@ -16,23 +16,159 @@
 #include <napa/log.h>
 
 #include <future>
+#include <unordered_map>
+#include <unordered_set>
 
 using namespace napa;
 using namespace napa::zone;
 
-// Static members initialization
-std::mutex NapaZone::_mutex;
-std::unordered_map<std::string, std::weak_ptr<NapaZone>> NapaZone::_zones;
+// A NapaZone has the following stage in its lifecycle:
+//  - Creating
+//      AFTER `zone.create()` is called and BEFORE all the workers are ready.
+//      * initialization stage *
+//      * zone JSObject is available to use *
+//      - `zone.broadcast()` and `zone.execute()` can be called but the tasks
+//        will not run until the next stage
+//
+//  - Created
+//      AFTER all the workers are ready and BEFORE either `zone.recycle()` is
+//      called or zone destructed triggered by GC.
+//      * the active stage *
+//
+//  - Recycling
+//      AFTER `zone.recycle()` is called and BEFORE zone destructed triggered by GC.
+//      * stop processing new request *
+//      * call to `zone.broadcast()` and `zone.execute()` should return a rejected
+//        Promise with error *
+//      - a zone may not enter this stage. If a zone is never called `zone.recycle()`
+//        manually, when it's destructed stage goes directly from 'Created' to
+//        'Recycled'.
+//
+//  - Recycled
+//      AFTER zone destructed triggered by GC and BEFORE all workers complete their
+//      remaining tasks
+//      * no active JSObject reference to this zone exists *
+//      * zone.get() returns null *
+//      * zone.current returns a recycled zone *
+//      - a zone may stay no time or infinite time in this stage, depending on what
+//        the rest tasks are in the workers' queue.
+//
+//  - Terminating
+//      AFTER all workers complete their remaining tasks and BEFORE all resources
+//      released
+//      - this stage should not wait for workers. Instead, it should be triggered by
+//        the signal that all workers complete their tasks.
+//
+//  - Terminated
+//      All resources should be released by this time point, including all C++/JS
+//      objects, all references of those and corresponding worker threads
+//
+// destructer may be triggered by napa_zone_release() via CAPI or javascript GC
```

Comment:
Personally I think we could simplify the concept. We should split Recycling/Recycled to other part only. Zone status could be Creating -- Created -- Stopping -- Stopped.
We could use a zone manager to hold reference to avoid directly action with recycle/recycled status.

I agree to check in this change first if it could work well.